docs/remote-attestation.md

```diff
@@ -0,0 +1,147 @@
+     bound to the TEE platform's firmware
+1. Data Exchange
+   - The keys generated during the Handshake are used to encrypt data sent
+     between Client and Server
+
+Prior to the _Handshake_ both Client and Server generate an individual pair of
+[ECDSA-P256](https://datatracker.ietf.org/doc/html/rfc6979) keys called
+**Signing Keys**. These keys are persistent between connections (but regenerated
+at startup) and are used to sign **Transcripts**: hashes of previous handshake
+messages that are sent in each new handshake message to prevent
+[Replay Attacks](https://en.wikipedia.org/wiki/Replay_attack). Each handshake
+message also includes a random string so that transcripts always contain new
+random values.
+
+Also for each individual connection both Client and Server generate a pair of
+[X25519](https://datatracker.ietf.org/doc/html/rfc7748) Diffie-Hellman keys
+called **Ephemeral Keys**. These keys are used to establish a shared secret
+between Client and Server.
+
+The shared secret is used to generate 2 shared
+[AES-256-GCM](https://datatracker.ietf.org/doc/html/rfc5288) **Session Keys**
+that are used during the _Data Exchange_ stage:
+
+- Server Session Key
+  - Which is used to encrypt messages sent by the Server
+- Client Session Key
+  - Which is used to encrypt messages sent by the Client
+
+## Workflow
+
+The workflow of the Remote Attestation protocol involves 3 interacting entities:
+
+1. Server
+   - Server application that processes requests from Clients
+   - Runs on a TEE Platform
+1. Client
+   - Client application that connects to the Server
+1. TEE Platform
+   - Firmware that provides a TEE support (i.e. Intel SGX or AMD-SEV-SNP capable
+     CPU)
+   - Possesses a firmware key
+1. TEE Provider
+   - TEE platform provider that can confirm authenticity of the TEE Platform
+     firmware keys using its root key
+   - Client must have the TEE Provider's root public key
+   - It’s important to note that the Provider is an external server (i.e.
+     belongs to Intel or AMD)
+
+The complete workflow of the Remote Attestation protocol looks as follows:
+
+1. Pre-handshake steps that are performed at startup:
+   - Both **Client** and **Server** generate an individual
+     [ECDSA-P256](https://datatracker.ietf.org/doc/html/rfc6979) _Signing_ key
+     pair
+   - **TEE Provider** creates a signature of the **TEE Platform**'s firmware key
+   - **Client** is provided with a **TEE Provider**'s _Root_ key
+1. **Client** generates an
+   [X25519](https://datatracker.ietf.org/doc/html/rfc7748) _Ephemeral_ key pair
+1. **Client** sends a `ClientHello` message to the **Server**
+   - Which includes a random string
+1. **Server** generates an
+   [X25519](https://datatracker.ietf.org/doc/html/rfc7748) _Ephemeral_ key pair
+1. **Server** sends a request for the `AttestationReport` to the **TEE
+   Platform**
+   - This request contains
+     [SHA-256](https://datatracker.ietf.org/doc/html/rfc6234) hash of the
```

Comment:
This request should probably also contain the client random string so the client can validate that it is was generated for this specific request.

Reply:
But I agree with @conradgrobler , if we don't pass the random string to the TEE, then the client cannot verify the freshness of the TEE signature, only that of the server, right?